### DISCUSSION TITLE
CDK lib code coverage is always 100%

### DISCUSSION THREAD

**Discussion by hopenbr** [aws/aws-cdk#23623] [Q&A]
We are building Infrastructure using CDK and we have unit tests via `Jest`.  The tests are helping us yet one issue we are seeing is that the code coverage is always 100%.  Any suggestion how on to get a more realistic view of code coverage?   

```shell 
=============================== Coverage summary ===============================
Statements   : 100% ( 73/73 )
Branches     : 100% ( 0/0 )
Functions    : 100% ( 10/10 )
Lines        : 100% ( 73/73 )
================================================================================

Test Suites: 10 passed, 10 total
Tests:       13 passed, 13 total
Snapshots:   0 total
Time:        112.165s

```

**Comment by khushail** [comment]
> We are building Infrastructure using CDK and we have unit tests via `Jest`. The tests are helping us yet one issue we are seeing is that the code coverage is always 100%. Any suggestion how on to get a more realistic view of code coverage?
> 
> ```shell
> =============================== Coverage summary ===============================
> Statements   : 100% ( 73/73 )
> Branches     : 100% ( 0/0 )
> Functions    : 100% ( 10/10 )
> Lines        : 100% ( 73/73 )
> ================================================================================
> 
> Test Suites: 10 passed, 10 total
> Tests:       13 passed, 13 total
> Snapshots:   0 total
> Time:        112.165s
> ```

@hopenbr , here are some suggestions on getting more realistic view of code coverage when using Jest -
1. Use a coverage tool like Istanbul - this will generate reports that show you which lines of code are covered by your tests and which are not.
2. write more tests.
3. Use a mocking framework like Jest's jest mock to mock out dependencies - this can help you to test code that depends on external resources such as databases or APIs.
4. Run your tests in Continuous intergation environment which will help you catch regression in code coverage as soon as it happens

Hope this is what you are asking for.

**Reply by khushail** [reply]
> > > We are building Infrastructure using CDK and we have unit tests via `Jest`. The tests are helping us yet one issue we are seeing is that the code coverage is always 100%. Any suggestion how on to get a more realistic view of code coverage?
> > > ```shell
> > > =============================== Coverage summary ===============================
> > > Statements   : 100% ( 73/73 )
> > > Branches     : 100% ( 0/0 )
> > > Functions    : 100% ( 10/10 )
> > > Lines        : 100% ( 73/73 )
> > > ================================================================================
> > > 
> > > Test Suites: 10 passed, 10 total
> > > Tests:       13 passed, 13 total
> > > Snapshots:   0 total
> > > Time:        112.165s
> > > ```
> > 
> > 
> > @hopenbr , here are some suggestions on getting more realistic view of code coverage when using Jest -
> > 
> > 1. Use a coverage tool like Istanbul - this will generate reports that show you which lines of code are covered by your tests and which are not.
> > 2. write more tests.
> > 3. Use a mocking framework like Jest's jest mock to mock out dependencies - this can help you to test code that depends on external resources such as databases or APIs.
> > 4. Run your tests in Continuous intergation environment which will help you catch regression in code coverage as soon as it happens
> > 
> > Hope this is what you are asking for.
> 
> Don't feel like you actually understand the problem (was this written by ChatGPT?). The problem is that test assertions are against the generated CloudFormation template, not against the code itself. So from the perspective of Jest, it's basically an integration test as it has to build the entire stack first before checking the assertions.

@hopenbr ,I was doubtful about understanding the question completely thats why I asked for a response. 

And I am not sure why you thought of this answer being generated by Chatgpt. I provided answer to the best of my knowledge. I am also on the path of understanding many aspects of CDK and testing so pardon me if my answer does not meet your requirements.

**Comment by renatomolina** [comment]
I wonder if the community has any suggestions on this issue. I can get any code coverage report working on the CDK project but not anything that compares the tests against the generated stack.

**Comment by mrpackethead** [comment]
This is a problem i'm facing as well.  I see that in the cdk-lib build process coverage reports.
